### PR TITLE
BCDA-350: Sign RPMs with GPG Keys

### PR DESCRIPTION
**Pull request by PatriciaAnong**
### Fixes [BCDA-350](https://jira.cms.gov/browse/BCDA-350)

Sign RPMs with GPG keys pulled as a secret from Jenkins
### Proposed changes:

- Update makefile to use the environment variables
- Update the dockerfile.package to pull the environment variables.
- Update the build and package script to sign the RPMs after they have been packaged.
- Update the jenkinsfie.build to pull the GPG secret keys and passphrase from Jenkins to be used by docker.

### Change Details

Updating the files listed above would allow enable the signng of the packaged RPMs for bcda and bcdaworker. The command used to sign the RPMs is 
```
echo $BCDA_GPG_RPM_PASSPHRASE | gpg --batch --yes --passphrase-fd 0 -v --pinentry-mode loopback --sign <filename>
```

### Security Implications

This allows us to verify the validity of the RPMs and easily detect any unauthorized changes/modifications to the packaged files.

### Acceptance Validation

Yes

image

### Feedback Requested

Verify that the changes are valid and the RPMs are indeed signed and valid.

### CODE REVIEW

File: Dockerfiles/Dockerfile.package [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
 
+ARG GPG_PUB_KEY_FILE
+ARG GPG_SEC_KEY_FILE
+ARG GPG_RPM_USER
+ARG GPG_RPM_EMAIL
+ARG WORKER_RPM
+ARG BCDA_RPM
+ARG BCDA_GPG_RPM_PASSPHRASE
+ARG GPG_PASSWORD_FILE
+
```
Comment (msnook) [inline Dockerfiles/Dockerfile.package]: Do you need all of those environment variables?  It looks like you are only passing in `BCDA_GPG_RPM_PASSPHRASE`.  All the other ones can probably be removed, right?

File: ops/build_and_package.sh [hunk 2]
```diff
@@ -33,3 +33,31 @@ go build
 echo "Packaging bcdaworker binary into RPM..."
 fpm -v $VERSION -s dir -t rpm -n bcdaworker bcdaworker=/usr/local/bin/bcdaworker
 
+#Sign RPMs
+GPG_RPM_USER="2AFAC6BABEFC07B6"
```
Comment (msnook) [inline ops/build_and_package.sh]: Is it acceptable to hardcode this?  Or should this be in Jenkins secrets?

File: ops/build_and_package.sh [hunk 3]
```diff
@@ -33,3 +33,31 @@ go build
...
+GPG_RPM_EMAIL="bcda-ops.group@adhocteam.us"
+GPG_PUB_KEY_FILE="../ops/RPM-GPG-KEY-bcda"
+GPG_SEC_KEY_FILE="../ops/RPM-GPG-KEY-SECRET-bcda"
+WORKER_RPM="bcdaworker-*.rpm"
```
Comment (msnook) [inline ops/build_and_package.sh]: Consider setting all constants at the top of the script for readability.

File: ops/build_and_package.sh [hunk 4]
```diff
@@ -33,3 +33,31 @@ go build
...
+/usr/bin/rpm --import $GPG_PUB_KEY_FILE
+
+echo "%_signature gpg %_gpg_path $PWD %_gpg_name $GPG_RPM_USER %_gpgbin /usr/bin/gpg" > $PWD/.rpmmacros
+echo "allow-loopback-pinentry" > ~/.gnupg/gpg-agent.conf
+
+echo "Signing bcdaworker RPM"
+echo $BCDA_GPG_RPM_PASSPHRASE | gpg --batch --yes --passphrase-fd 0 -v --pinentry-mode loopback --sign $WORKER_RPM
+
+cd ../bcda
+BCDA_RPM="bcda-*.rpm"
```
Comment (msnook) [inline ops/build_and_package.sh]: Please this constant with the rest of the constants above.

File: ops/build_and_package.sh (hunk 3 above)
Comment (msnook) [reply ops/build_and_package.sh]: Please add a check at the top of the script which fails execution if any of the env vars are not set.

**Comment by PatriciaAnong** [comment]
Made the suggested changes and ready for another review

**Pull request closed without merging**